### PR TITLE
Fix DoQ default configuration

### LINKED ISSUE #1990: DoQ default configuration is invalid
Currently DoT, DoH and DoQ share the same `ClientConfig`:
https://github.com/bluejekyll/trust-dns/blob/bf04bc8d3efec41d29cf558416b3dbed2d26ae02/crates/resolver/src/tls/dns_over_rustls.rs#L30-L56

All three implementations can't share the same configuration as they differ on some properties:
- **ALPN** (mandatory)
  - DoT: None
  - DoH: "h2" (current)
  - DoQ: "doq"
- **SNI** (optional)
  We probably want to continue disable SNI for DoT by default (though the only public server that I found supporting DoQ, AdGuard, requires SNI for both DoQ and DoT), but disabling SNI for DoH by default is strange.

The fast fix is probably easy, just use a different `ClientConfig` for each protocol. This does have a big downside though, when using multiple backends this will parse the root certificates multiple times.

This in turn could be solved by using [`WebPkiVerifier`](https://docs.rs/rustls/0.21.5/rustls/client/struct.WebPkiVerifier.html), but that would require `dangerous_configuration` and I'm not sure if we want to use that because it would activate it for upstream users as well.

So unless we are fine with activating `dangerous_configuration`, this would probably require some fix in Rustls. Considering @djc maintains both, I'm gonna hold back from creating an issue there until further feedback.

### PR DESCRIPTION
**Pull request by daxpedda**
It has become clear to me that #2031 and #2003 will require more design work.
So until we figure out what to do there I thought I can tackle some of the things we already discussed with what we have in place right now.

Fixes #1990.

### CODE REVIEW

File: crates/resolver/src/tls/dns_over_rustls.rs [hunk 1]
```diff
@@ -72,8 +70,6 @@ pub(crate) static CLIENT_CONFIG: Lazy<Result<Arc<ClientConfig>, ProtoError>> = L
     // The port (853) of DOT is for dns dedicated, SNI is unnecessary. (ISP block by the SNI name)
     client_config.enable_sni = false;
 
-    client_config.alpn_protocols.push(ALPN_H2.to_vec());
```
Comment (bluejekyll) [inline crates/resolver/src/tls/dns_over_rustls.rs:75]: hm, I think we might have been doing the same thing in the server, https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/rustls/tls_server.rs#L103

I'm not sure why I did that, but I think we'll need to thread that through properly for the DoH settings?

Comment (daxpedda) [reply crates/resolver/src/tls/dns_over_rustls.rs:75]: > hm, I think we might have been doing the same thing in the server, https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/rustls/tls_server.rs#L103

I can look into it, should I do this here or in a different PR?
> I'm not sure why I did that, but I think we'll need to thread that through properly for the DoH settings?

The client already inserts the right ALPN when it finds that none were specified in all protocols so far:
https://github.com/bluejekyll/trust-dns/blob/90edda6b0558282abf852494d7c4f4004e47e1d3/crates/proto/src/https/https_client_stream.rs#L311-L316
https://github.com/bluejekyll/trust-dns/blob/90edda6b0558282abf852494d7c4f4004e47e1d3/crates/proto/src/quic/quic_client_stream.rs#L245-L247

Comment (djc) [reply crates/resolver/src/tls/dns_over_rustls.rs:75]: So this merely removes the H2 ALPN from the client config. Doesn't this negatively impact the H2 setup? What do DoH (2) standards say on the use of ALPN?

Comment (daxpedda) [reply crates/resolver/src/tls/dns_over_rustls.rs:75]: > Doesn't this negatively impact the H2 setup?

It has no impact on the H2 setup.
https://github.com/bluejekyll/trust-dns/blob/90edda6b0558282abf852494d7c4f4004e47e1d3/crates/proto/src/https/https_client_stream.rs#L311-L316
This part of the code inserts the required ALPN when no ALPN was found. Removing the initial DoH ALPN was necessary to make other protocols which require different ALPNs work, namely DoQ and DoH3.

> What do DoH (2) standards say on the use of ALPN?

I don't believe DoH has anything to say about ALPN, but HTTP/2 [requires it](https://www.rfc-editor.org/rfc/rfc9113.html#name-http-2-version-identificati).

Comment (bluejekyll) [reply crates/resolver/src/tls/dns_over_rustls.rs:75]: I think that's right, though DoH only works with HTTP/2, so for DoH we should be setting it. Just not for TLS.

**Comment by daxpedda** [comment]
Rebased again.

File: crates/resolver/src/tls/dns_over_rustls.rs (hunk 1 above)
Comment (daxpedda) [reply crates/resolver/src/tls/dns_over_rustls.rs:75]: > [..], so for DoH we should be setting it. Just not for TLS.

Great! That's what we are doing at the moment.
Gonna mark this as resolved then.

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
(Sorry for the slow response -- busy week.)

**Pull request merged**